daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala

```diff
@@ -89,156 +97,190 @@ private[engine] final class ValueTranslator(compiledPackages: CompiledPackages)
+                    if (entries.isEmpty) {
+                      SValue.SValue.EmptyTextMap
+                    } else {
+                      SValue.SGenMap(
+                        isTextMap = false,
+                        entries = entries.iterator.map { case (k, v) =>
+                          SValue.SText(k) -> go(typeArg0, v, newNesting)
+                        },
+                      )
+                    }
+                  case _ =>
+                    typeMismatch
+                }
+              case typeArg0 :: typeArg1 :: Nil =>
+                (bt, value) match {
+                  case (BTGenMap, ValueGenMap(entries)) =>
+                    if (entries.isEmpty) {
+                      SValue.SValue.EmptyTextMap
+                    } else {
+                      SValue.SGenMap(
+                        isTextMap = true,
+                        entries = entries.iterator.map { case (k, v) =>
+                          go(typeArg0, k, newNesting) -> go(typeArg1, v, newNesting)
+                        },
+                      )
+                    }
+                  case _ =>
+                    typeMismatch
+                }
+              case _ =>
+                typeMismatch
+            }
+          case TTyCon(tyCon) =>
+            val pkg = unsafeGetPackage(tyCon.packageId)
+            value match {
+              // variant
+              case ValueVariant(mbId, constructorName, val0) =>
+                mbId.foreach(id =>
+                  if (id != tyCon)
+                    fail(
+                      s"Mismatching variant id, the type tells us $tyCon, but the value tells us $id"
+                    )
                 )
-              case Some(rank) =>
-                val (_, argTyp) = variantDef.variants(rank)
-                val instantiatedArgTyp =
-                  replaceParameters(dataTypParams.map(_._1).zip(tyConArgs), argTyp)
-                SValue.SVariant(
-                  typeVariantId,
-                  constructorName,
-                  rank,
-                  go(instantiatedArgTyp, val0, newNesting),
+                val (dataTypParams, variantDef) = assertRight(
+                  SignatureLookup.lookupVariant(pkg, tyCon.qualifiedName)
                 )
-            }
-          // records
-          case (TTyConApp(typeRecordId, tyConArgs), ValueRecord(mbId, flds)) =>
-            mbId.foreach(id =>
-              if (id != typeRecordId)
-                fail(
-                  s"Mismatching record id, the type tells us $typeRecordId, but the value tells us $id"
+                variantDef.constructorRank.get(constructorName) match {
+                  case None =>
+                    fail(
+                      s"Couldn't find provided variant constructor $constructorName in variant $tyCon"
+                    )
+                  case Some(rank) =>
+                    val (_, argTyp) = variantDef.variants(rank)
+                    val replacedTyp =
```

Comment:
is `toSeq.view` actually more efficient than just `map`?

Reply:
In this case, I think it is.

First I need the result to be `Iterable` (to be able to use `.toMap`). 

1 - Since `dataTypParams` is an ImmArray I need a `.toSeq` somewhere. 
2 - `dataTypParams.toSeq.map(_._1).zip(tyArgs)` will allocate 2 intermediate `ImmArraySeq`
3 - In general Array builder allocate several intermediate arrays (as they do not know the final size of the array they are building).